sequencer.go

```diff
@@ -1,42 +1,94 @@
 package desync
 
+import "sort"
+
 // SeedSequencer is used to find sequences of chunks from seed files when assembling
 // a file from an index. Using seeds reduces the need to download and decompress chunks
 // from chunk stores. It also enables the use of reflinking/cloning of sections of
 // files from a seed file where supported to reduce disk usage.
 type SeedSequencer struct {
-	seeds   []Seed
-	index   Index
-	current int
+	seeds             []Seed
+	index             Index
+	invalidSeedAction InvalidSeedAction
+	current           int
+}
+
+type matchingCandidate struct {
+	seed    Seed
+	source  SeedSegment
+	advance int
+}
+
+type matchingCandidateList []matchingCandidate
+
+func (m matchingCandidateList) Len() int {
+	return len(m)
+}
+
+func (m matchingCandidateList) Less(i, j int) bool {
+	// Sort the candidates in descending order based on the segment size
+	return m[i].source.Size() > m[j].source.Size()
+}
+
+func (m matchingCandidateList) Swap(i, j int) {
+	m[i], m[j] = m[j], m[i]
 }
 
 // NewSeedSequencer initializes a new sequencer from a number of seeds.
-func NewSeedSequencer(idx Index, src ...Seed) *SeedSequencer {
+func NewSeedSequencer(idx Index, invalidSeedAction InvalidSeedAction, src ...Seed) *SeedSequencer {
 	return &SeedSequencer{
-		seeds: src,
-		index: idx,
+		seeds:             src,
+		index:             idx,
+		invalidSeedAction: invalidSeedAction,
 	}
 }
 
 // Next returns a sequence of index chunks (from the target index) and the
 // longest matching segment from one of the seeds. If source is nil, no
 // match was found in the seeds and the chunk needs to be retrieved from a
 // store. If done is true, the sequencer is complete.
-func (r *SeedSequencer) Next() (segment IndexSegment, source SeedSegment, done bool) {
+func (r *SeedSequencer) Next() (segment IndexSegment, source SeedSegment, done bool, err error) {
 	var (
-		max     uint64
-		advance = 1
+		advance   = 1
+		candidate []matchingCandidate
 	)
 	for _, s := range r.seeds {
+		// If we already know that this seed is invalid, we just skip it
+		if s.IsInvalid() {
+			continue
+		}
 		n, m := s.LongestMatchWith(r.index.Chunks[r.current:])
-		if n > 0 && m.Size() > max {
-			source = m
-			advance = n
-			max = m.Size()
+
+		if n > 0 {
+			candidate = append(candidate, matchingCandidate{s, m, n})
+		}
+	}
+
+	// Sort the candidates from the longest to the shortest
+	sort.Sort(matchingCandidateList(candidate))
+	for _, matching := range candidate {
+		err := matching.source.Validate()
```

Comment:
Thinking about it a bit more I remember now why I didn't do it this way originally and delayed the validation. This code here runs in a single goroutine, while the copying of data happens concurrently. So previously we had N goroutines validating and copying the seed chunks. Now we validate in a single one, but still write concurrently. This should indeed be slower in that case. While I still like this change, we may have to add the optimization to remember validated chunks in order to get at least some of the performance back.

Reply:
Ah I see. Well this could really negatively impact the performance.

I'm thinking out loud, what would happen if we added Goroutine in `Validate()` too? We could split the segment in X parts and give each one of them to different Goroutine.

Reply:
That is certainly an option. If the candidate has multiple chunks, each of them could be validated concurrently with up to N goroutines. Would have to use the same filehandle as opening new ones per chunk would be too expensive. I suspect this will get back some of the performance, but perhaps not all of it. Combined with a way to remember which chunks have been validated it may get close.

A more efficient and more complex way to do this would be to introduce the concept of a "plan" for how to assemble the file. Could plan it out fully without validation, then run a validation step on the whole plan (concurrently), if a bad seed is found we throw away the plan (and mark the seed bad), and re-run the plan. That what I wanted to do to deal with this originally but never got around to.

Reply:
I can help with that part if you like. And when combining that with remembering what chunks are validated in a seed, this would make the whole process faster than before

Reply:
Thanks, I guess that your suggestion to use a "plan" is clear enough. I should be able to propose a first implementation "soon", and then we can iterate/improve from that.

Reply:
I added the concept of the "plan" in my latest revision. I run a couple of quick benchmarks and the difference between this patch and the current master seems to be slim (master ~6m38s and with this patch ~6m55s).

I didn't include the part about remembering which chunks were already validated to avoid complicating too much this PR. But I think that this is definitely something that we should do.